### PR TITLE
Fix building error in expo project

### LINKED ISSUE #1237: cannot find installWatermelonJSI & watermelondbProvideSyncJson in scope
My source code build ios app on MacOS chip M1: **success** 
But **fail** if build it on MacOS chip Intel x86_64

both same:
- MacOS 12.1
- Xcode: 13.2.1
---- and ----
- MacOS 12.0
- Xcode: 13.2.0

```
› Compiling @nozbe/watermelondb Pods/WatermelonDB » std_ext.swift
› Compiling @nozbe/watermelondb Pods/WatermelonDB » DatabaseBridge.swift

❌  (node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/DatabaseBridge.swift:93:13)

  91 |         methodQueue.sync {
  92 |             // swiftlint:disable all
> 93 |             installWatermelonJSI(bridge as? RCTCxxBridge)
     |             ^ cannot find 'installWatermelonJSI' in scope
  94 |         }
  95 |         return [:]
  96 |     }


❌  (node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/DatabaseBridge.swift:103:9)

  101 |                          resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
  102 |         var error: NSError?
> 103 |         watermelondbProvideSyncJson(id.int32Value, json.data(using: String.Encoding.utf8.rawValue), &error)
      |         ^ cannot find 'watermelondbProvideSyncJson' in scope
  104 |         if let error = error {
  105 |             sendReject(reject, error)
  106 |         } else {

› Compiling @nozbe/watermelondb Pods/WatermelonDB » Database.swift
› Compiling @nozbe/watermelondb Pods/WatermelonDB » DatabaseDriver.swift
```

My package:
```
"@nozbe/watermelondb": "~0.24.0",
"expo": "^44.0.0",
"react-native": "0.64.3",
```

I don't know why? Please help me.

Thanks so much

### PR DESCRIPTION
**Pull request by Kudo**
# Why

in expo sdk 44+ project, we opened formal swift integration. for swift to call objective-c functions, sometimes the `#import` format should be done right. learn more from https://github.com/expo/expo/issues/15622#issuecomment-997225774

fix #1237 
fix https://github.com/expo/expo/issues/15986
 
# How

watermelonDB has a call from `DatabaseBridge.swift` to `JSIInstaller.h`. for swift to call objective-c functions, we should make `JSIInstaller.h` as a clang module. this pr moves `JSIInstaller.h` as cocoapods public header and the generated umbrella header will include `JSIInstaller.h` as a clang submodule.

# Test Plan

### expo sdk 44 project

1. expo init sdk44 # select bare
2. yarn add @nozbe/watermelondb
3. patch watermelondb with the pr changes
4. add these to `ios/Podfile`
```
  pod 'React-jsi', :path => '../node_modules/react-native/ReactCommon/jsi', :modular_headers => true
  pod 'simdjson', path: '../node_modules/@nozbe/simdjson'
```
4. npx pod-install
5. yarn ios

### react-native 0.66 project

1. npx react-native init RN066 --version 0.66
2. yarn add @nozbe/watermelondb
3. patch watermelondb with the pr changes
4. add these to `ios/Podfile`
```
  pod 'React-jsi', :path => '../node_modules/react-native/ReactCommon/jsi', :modular_headers => true
  pod 'simdjson', path: '../node_modules/@nozbe/simdjson'
```
4. npx pod-install
5. yarn ios

### CODE REVIEW

**Comment by enahum** [comment]
@radex thoughts on having this merged ?

**Comment by aircliff** [comment]
Hey All, any reason this PR hasnt been merged? WatermelonDB is broken and pretty unusable on expo 44. This PR fixes the issue.

**Comment by thecoorum** [comment]
@Kudo does this work for expo projects with `dev-client` and `expo prebuild -p ios`? After following instructions it's failing on EAS Fastlane step with:
```
Compiling @nozbe/watermelondb Pods/WatermelonDB » Database.swift

❌  (node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/DatabaseBridge.swift:93:13)

  91 |         methodQueue.sync {
  92 |             // swiftlint:disable all
> 93 |             installWatermelonJSI(bridge as? RCTCxxBridge)
     |             ^ cannot find 'installWatermelonJSI' in scope
  94 |         }
  95 |         return [:]
  96 |     }

❌  (node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/DatabaseBridge.swift:103:9)

  101 |                          resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
  102 |         var error: NSError?
> 103 |         watermelondbProvideSyncJson(id.int32Value, json.data(using: String.Encoding.utf8.rawValue), &error)
      |         ^ cannot find 'watermelondbProvideSyncJson' in scope
  104 |         if let error = error {
  105 |             sendReject(reject, error)
  106 |         } else {

⚠️  (../../Library/Developer/Xcode/DerivedData/track-ezmllxjvzftvvzcdtnxcfnrldxbo/Build/Intermediates.noindex/ArchiveIntermediates/track/IntermediateBuildFilesPath/Pods.build/Debug-iphoneos/WatermelonDB.build/Objects-normal/arm64/Database-f1a153e176371a1f77b80228b6a743e6.dia:1:1)

> 1 | 
    | ^ Could not read serialized diagnostics file: error(in target 'WatermelonDB' from project 'Pods')
› Compiling @nozbe/watermelondb Pods/WatermelonDB » DatabaseBridge.swift

❌  (node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/DatabaseBridge.swift:93:13)

  91 |         methodQueue.sync {
  92 |             // swiftlint:disable all
> 93 |             installWatermelonJSI(bridge as? RCTCxxBridge)
     |             ^ cannot find 'installWatermelonJSI' in scope
  94 |         }
  95 |         return [:]
  96 |     }

❌  (node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/DatabaseBridge.swift:103:9)

  101 |                          resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
  102 |         var error: NSError?
> 103 |         watermelondbProvideSyncJson(id.int32Value, json.data(using: String.Encoding.utf8.rawValue), &error)
      |         ^ cannot find 'watermelondbProvideSyncJson' in scope
  104 |         if let error = error {
  105 |             sendReject(reject, error)
  106 |         } else {
```
and
```
The following build commands failed:
▸ 	CompileSwift normal arm64 /Users/expo/workingdir/build/node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/Database.swift (in target 'WatermelonDB' from project 'Pods')
▸ 	CompileSwift normal arm64 /Users/expo/workingdir/build/node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/DatabaseBridge.swift (in target 'WatermelonDB' from project 'Pods')
▸ 	CompileSwiftSources normal arm64 com.apple.xcode.tools.swift.compiler (in target 'WatermelonDB' from project 'Pods')
▸ 	CompileSwiftSources normal arm64 com.apple.xcode.tools.swift.compiler (in target 'expo-dev-menu' from project 'Pods')
▸ (4 failures)
```

**Comment by Kudo** [comment]
@thecoorum do you have some example repo? because the installation requires add extra pods to `Podfile`, i am assuming you would have a bare workflow app. otherwise, a config-plugin to add pods is required.
i've verified with this steps and it works for me.

1. expo init sdk44 # select bare
2. expo install expo-dev-client
3. yarn add @nozbe/watermelondb
4. patch watermelondb with the pr changes
5. add these to `ios/Podfile`
```
  pod 'React-jsi', :path => '../node_modules/react-native/ReactCommon/jsi', :modular_headers => true
  pod 'simdjson', path: '../node_modules/@nozbe/simdjson'
```
4. npx pod-install
6. yarn ios

**Comment by KrisLau** [comment]
@Kudo Waiting on this to be merged as migrating to expo Modules (specifically Expo 44) from unimodules breaks WatermelonDB

**Review by ruifernandees — APPROVED** [review]
I've tested and it is working good.

**Review by vtpa — APPROVED** [review]
Approved!

**Comment by zsinryu** [comment]
@vtpa @Kudo this have been approved.
Can you guys please merge this?

**Comment by jwoo92** [comment]
I was able to get `@nozbe/watermelondb@0.24.0` to work with Expo 44 after applying this patch.

Thank you 👍

**Comment by KrisLau** [comment]
> @vtpa @Kudo this have been approved. Can you guys please merge this?

@zsinryu They cannot merge the pull request since I believe only @radex has that permission

**Comment by Hannes1** [comment]
Hey guys, can someone please post their patch package diff file, mine keeps failing when I generate it😭

**Comment by jwoo92** [comment]
> Hey guys, can someone please post their patch package diff file, mine keeps failing when I generate it😭

https://github.com/Nozbe/WatermelonDB/issues/1237#issuecomment-1114305677

**Comment by Hannes1** [comment]
> > Hey guys, can someone please post their patch package diff file, mine keeps failing when I generate it😭
> 
> https://github.com/Nozbe/WatermelonDB/issues/1237#issuecomment-1114305677

Thanks, I tried this already and keep getting a error when running npx patch, this is why I'm asking for the file someone already generated in their patch folder

**Comment by jwoo92** [comment]
@Hannes1 Oh I missed that, here you go:

`patches/@nozbe+watermelondb+0.24.0.patch`

```
diff --git a/node_modules/@nozbe/watermelondb/WatermelonDB.podspec b/node_modules/@nozbe/watermelondb/WatermelonDB.podspec
index 1f3af50..e9fea58 100644
--- a/node_modules/@nozbe/watermelondb/WatermelonDB.podspec
+++ b/node_modules/@nozbe/watermelondb/WatermelonDB.podspec
@@ -13,7 +13,10 @@ Pod::Spec.new do |s|
   s.platforms    = { :ios => "9.0", :tvos => "9.0" }
   s.source = { :git => "https://github.com/Nozbe/WatermelonDB.git", :tag => "v#{s.version}" }
   s.source_files = "native/ios/**/*.{h,m,mm,swift,c,cpp}", "native/shared/**/*.{h,c,cpp}"
-  s.public_header_files = 'native/ios/WatermelonDB/SupportingFiles/Bridging.h'
+  s.public_header_files = [
+    'native/ios/WatermelonDB/SupportingFiles/Bridging.h',
+    'native/ios/WatermelonDB/JSIInstaller.h',
+  ]
   s.requires_arc = true
   # simdjson is annoyingly slow without compiler optimization, disable for debugging
   s.compiler_flags = '-Os'
diff --git a/node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/SupportingFiles/Bridging.h b/node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/SupportingFiles/Bridging.h
index 135118d..8324550 100644
--- a/node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/SupportingFiles/Bridging.h
+++ b/node_modules/@nozbe/watermelondb/native/ios/WatermelonDB/SupportingFiles/Bridging.h
@@ -6,12 +6,6 @@
 
 #import <React/RCTBridgeModule.h>
 
-#if __has_include("JSIInstaller.h")
-#import "JSIInstaller.h"
-#else
-#import "../JSIInstaller.h"
-#endif
-
 #if __has_include("DatabaseDeleteHelper.h")
 #import "DatabaseDeleteHelper.h"
 #else
```

**Comment by Hannes1** [comment]
@jwoo92 Greatly appreciated, thanks for the help🙏

**Review by diegoalvescf — APPROVED** [review]
Approved, I want this PR until tomorrow merged

**Review by jamesjlv — APPROVED** [review]
Approved!

**Comment by zsinryu** [comment]
@radex  why this is not merged?

**Comment by alexyoungs** [comment]
Hey guys, checking in on whether this can be merged?

**Comment by KrisLau** [comment]
@alexyoungs Read the above thread. None of us have permission to do so, so it's currently just awaiting review/approval from @radex

**Comment by ansh** [comment]
Why is this PR not merged? It seems like it works.

Pining @radex and @RafalNozbe for help

**Comment by ItzzRitik** [comment]
@radex This PR needs to be merged, as WatermelonDB is no longer working for EXPO 44+
Looks like it's working for most people with this fix.

Can we merge it?

**Comment by KrisLau** [comment]
@Kudo Hi sorry to bother but I just noticed a related issue now that I'm doing an update to some of my packages. Whenever I add `use_frameworks! :linkage => :static` to my Podfile (to work with RNFirebase), I can fix the error from the podspec by using the fixes mentioned in  get a number of errors related to the `Bridging.h` file:

image

Not meaning to hijack but I just noticed that `expo-modules-core` has some fixes for this in their heading file:
image
I'm trying to implement something similar to get the `use_frameworks!` to work in my project but I don't have enough understanding of swift to know how to implement this. Would you be able to help?

Created issue: #1436

**Comment by ansh** [comment]
@radex Some communication is the bare minimum after months of a valid PR from the Expo team.

**Comment by radex** [comment]
@Kudo Thank you

**Comment by Dallas62** [comment]
Hi @radex, I still had to change the content of `WatermelonDB.podspec` with: 

```pod
  s.public_header_files = [
      'native/ios/WatermelonDB/**/*.h',
  ]
```

And copy :
```
node_modules/@nozbe/simdjson/src/simdjson.h
```
In :
```
node_modules/@nozbe/watermelondb/native/shared
```

On 0.70.6

**Comment by radex** [comment]
@Dallas62 Can you please elaborate? What issues did you see, what changes did you have to make, which Expo version?

**Comment by Dallas62** [comment]
This issue doesn't happen with Expo, it's mainly not resolve import in the native side (same file of this issue).

Will provide more details later

**Comment by radex** [comment]
@Dallas62 Please do provide more details. Are you using  Turbo Sync by any chance?

**Comment by KrisLau** [comment]
@radex so the issue @Dallas62 was facing is probably related to `use_frameworks!` and I also have the same thing patched. It would be nice to be able to have it permanently fixed since[ react-native-firebase requires the use of `use_frameworks!` now](https://github.com/Nozbe/WatermelonDB/issues/1285#issuecomment-1381323060).

Relevant thread: https://github.com/Nozbe/WatermelonDB/issues/1285#issuecomment-1381323060

My patch including this PR
```

```

**Pull request merged**
